### PR TITLE
refactor connector

### PR DESCRIPTION
**Pull request by jshencode**
extract schema and enum logic
SchemaFetcher is pluggable

### CODE REVIEW

File: client/schema.go [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+	resp, err := hf.httpClient.Get(hf.tablePath(tableName))
+	err = hf.readJSONResponse(resp, err, &table)
+	if err != nil {
+		return nil, err
+	}
+
+	return &table, nil
+}
+
+func (hf *httpSchemaFetcher) FetchAllSchemas() ([]*metaCom.Table, error) {
```
Comment (voyager-dw) [inline client/schema.go]: possible to combine into one http call?

File: client/schema.go [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+		if caseInsensitive {
+			enumCase = strings.ToLower(enumCase)
+		}
+		cf.enumMappings[tableName][columnID][enumCase] = enumIDs[index]
+	}
+	cf.Unlock()
+	return nil
+}
+
+func (cf *CachedSchemaHandler) fetchAndSetEnumCases(table *metaCom.Table) error {
```
Comment (voyager-dw) [inline client/schema.go]: also, possible to combine to fetch all column enums in one http call?

File: client/connector.go [hunk 3]
```diff
@@ -527,6 +349,12 @@ func (u *UpsertBatchBuilderImpl) PrepareUpsertBatch(tableName string, columnName
 		if err = upsertBatchBuilder.AddColumnWithUpdateMode(columnID, dataType, updateModes[colIndex]); err != nil {
 			return nil, 0, err
 		}
+
+		if column.IsEnumColumn() {
+			if err = u.prepareEnumCases(tableName, columnName, colIndex, columnID, rows, abandonRows, column.CaseInsensitive, column.DisableAutoExpand); err != nil {
```
Comment (voyager-dw) [inline client/connector.go]: this maybe an overhead for most no need to extend cases.

File: client/connector.go [hunk 4]
```diff
@@ -128,43 +112,28 @@ func (cfg ConnectorConfig) NewConnector(logger *zap.SugaredLogger, metricScope t
...
-		UpsertBatchBuilderImpl: &upsertBatchBuilderImpl,
+	httpClient := http.Client{
+		Timeout: time.Duration(cfg.Timeout) * time.Second,
 	}
 
-	connector.initHTTPClient()
-
-	err := connector.fetchAllTables()
+	httpSchemaFetcher := NewHttpSchemaFetcher(httpClient, cfg.Address, metricScope)
+	cachedSchemaHandler := NewCachedSchemaHandler(logger, metricScope, httpSchemaFetcher)
```
Comment (voyager-dw) [inline client/connector.go]: nice wrapper

File: client/connector.go (hunk 3 above)
Comment (jshencode) [reply client/connector.go]: this will compare with local cache first. only new cases will be extended

File: client/schema.go (hunk 2 above)
Comment (jshencode) [reply client/schema.go]: maybe later? currently we do not support combined call in the api. plus this is not really a performance bottleneck.

File: client/schema.go (hunk 1 above)
Comment (jshencode) [reply client/schema.go]: same here. maybe later when we see necessities for this change

**Pull request merged**